File: pset6/sentimental-readability/readability.py
from cs50 import get_string
import re

# function to count words in a string


def words(string):
    to_words = string.split()
    return len(to_words)


# function to count letters in a string


def letters(string):
    to_letters = [char for char in string if char.isalpha()]
    return len(to_letters)


# function to count sentences in a string


def sentences(string):
    to_sentences = [s for s in re.split(r'[.!?]', string) if s.strip(' "')]
    return len(to_sentences)


# main


def main():

    # get text from user
    text = get_string("Text: ")

    # Calculate letters, words and sentences
    le = letters(text)
    wo = words(text)
    se = sentences(text)

    # Calculate averages
    L = 100 * (le / wo)
    S = 100 * (se / wo)

    # Calculate Coleman-Liau index
    coli_index = round(0.0588 * L - 0.296 * S - 15.8)

    # Print Resulting Grade
    if coli_index >= 1 and coli_index < 16:
        print(f"Grade {coli_index}")
    elif coli_index >= 16:
        print(f"Grade 16+")
    elif coli_index < 1:
        print("Before Grade 1")


# run main program
if __name__ == "__main__":
    main()

File: pset2/readability/readability.c
#include <cs50.h>
#include <ctype.h>
#include <math.h>
#include <stdio.h>
#include <string.h>

int compute_grade(string);

// calculates approximate grade level for a given text
int main(void)
{
    // get text from user
    string text = get_string("Text: ");

    // compute grade level
    int grade_level = compute_grade(text);

    // print the appropriate message
    if (grade_level < 16 && grade_level > 1)
    {
        printf("Grade %i\n", grade_level);
    }
    else if (grade_level < 1)
    {
        printf("Before Grade 1\n");
    }
    else
    {
        printf("Grade 16+\n");
    }
}

// implement compute_grade
int compute_grade(string text)
{
    // calculating the number of letters in text
    int letters = 0;
    for (int i = 0, len = strlen(text); i < len; i++)
    {
        if (isalpha(text[i]))
        {
            letters++;
        }
    }

    // calculating number of spaces
    int spaces = 0;
    for (int i = 0, len = strlen(text); i < len; i++)
    {
        if (text[i] == ' ')
        {
            spaces++;
        }
    }

    // number of words
    int words = spaces + 1;

    // number of sentences
    int sentences = 0;
    for (int i = 0, len = strlen(text); i < len; i++)
    {
        if (text[i] == '.' || text[i] == '?' || text[i] == '!')
        {
            sentences++;
        }
    }
    float L = 100 * ((float) letters / (float) words);
    float S = 100 * ((float) sentences / (float) words);
    int index = round(0.0588 * L - 0.296 * S - 15.8);
    return index;
}
